flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -1116,148 +1115,177 @@ private void restoreKeyGroupsShardWithTemporaryHelperInstance(
 	// ------------------------------------------------------------------------
 
 	/**
-	 * Creates a column family handle for use with a k/v state. When restoring from a snapshot
-	 * we don't restore the individual k/v states, just the global RocksDB database and the
-	 * list of column families. When a k/v state is first requested we check here whether we
-	 * already have a column family for that and return it or create a new one if it doesn't exist.
+	 * Registers a k/v state information, which includes its state id, type, RocksDB column family handle, and serializers.
 	 *
-	 * <p>This also checks whether the {@link StateDescriptor} for a state matches the one
-	 * that we checkpointed, i.e. is already in the map of column families.
+	 * When restoring from a snapshot, we don’t restore the individual k/v states, just the global RocksDB database and
+	 * the list of k/v state information. When a k/v state is first requested we check here whether we
+	 * already have a registered entry for that and return it (after some necessary state compatibility checks)
+	 * or create a new one if it does not exist.
 	 */
-	@SuppressWarnings("rawtypes, unchecked")
-	protected <N, S> ColumnFamilyHandle getColumnFamily(
-		StateDescriptor<?, S> descriptor, TypeSerializer<N> namespaceSerializer) throws IOException, StateMigrationException {
+	private Tuple2<ColumnFamilyHandle, RegisteredKeyedBackendStateMetaInfo<?, ?>> tryRegisterKvStateInformation(
```

Comment:
We could rewrite this as
```
	private <N, S> Tuple2<ColumnFamilyHandle, RegisteredKeyedBackendStateMetaInfo<N, S>> tryRegisterKvStateInformation(
			StateDescriptor<?, S> stateDesc,
			TypeSerializer<N> namespaceSerializer) throws StateMigrationException, IOException {

		Tuple2<ColumnFamilyHandle, RegisteredKeyedBackendStateMetaInfo<?, ?>> registeredInfo =
			this.kvStateInformation.get(stateDesc.getName());

		if (registeredInfo != null) {

			@SuppressWarnings("unchecked")
			RegisteredKeyedBackendStateMetaInfo.Snapshot<?, ?> restoredMetaInfoSnapshot =
				restoredKvStateMetaInfos.get(stateDesc.getName());

			Preconditions.checkState(
				restoredMetaInfoSnapshot != null,
				"Requested to check compatibility of a restored RegisteredKeyedBackendStateMetaInfo," +
					" but its corresponding restored snapshot cannot be found.");

			RegisteredKeyedBackendStateMetaInfo<N, S> resolveKvStateCompatibility =
				RegisteredKeyedBackendStateMetaInfo.resolveKvStateCompatibility(
				restoredMetaInfoSnapshot,
				namespaceSerializer,
				stateDesc);

			registeredInfo.f1 = resolveKvStateCompatibility;

			return Tuple2.of(registeredInfo.f0, resolveKvStateCompatibility);
		} else {
			String stateName = stateDesc.getName();
			RegisteredKeyedBackendStateMetaInfo<N, S> newMetaInfo = new RegisteredKeyedBackendStateMetaInfo<>(
				stateDesc.getType(),
				stateName,
				namespaceSerializer,
				stateDesc.getSerializer());

			ColumnFamilyHandle columnFamily = createColumnFamily(stateName);
			registeredInfo = Tuple2.of(columnFamily, newMetaInfo);
			this.kvStateInformation.put(stateDesc.getName(), registeredInfo);
			return Tuple2.of(columnFamily, newMetaInfo);
		}
	}
```

and get rid of all the individual casts.

Reply:
And we can remove also the lines that suppress warnings

Reply:
I did not go with this approach because of the extra tuples introduced.
Though, TBH, I wasn't sure which was the better approach, this or the one you pointed out.

I would not be against this version.

Reply:
Will change this as suggested!